ENH: Enabled skipna argument on groupby reduction ops #15675

andremcorreia opened the pull request:
- [x] closes #15675
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst` file if fixing a bug or adding a new feature.

Added a skipna argurment to the groupby reduction ops for consistency with the Series and Dataframe variants:
- sum, 
- prod, 
- min, 
- max, 
- mean, 
- median, 
- var, 
- std,
- sem

Added new relevant tests, updated api tests and whatsnew

andremcorreia committed:
ENH: Enabled skipna argument on groupby reduction ops (#15675)

Added a skipna argurment to the groupby reduction ops:
  sum, prod, min, max, mean, median, var, std and sem
Added relevant tests
Updated whatsnew to reflect changes

Co-authored-by: Tiago Firmino <tiago.esteves.firmino@tecnico.ulisboa.pt>

tiago-firmino committed:
FIX: fixed pipeline issues related to docs and window tests

Co-authored-by: André Correia <andre.m.correia@tecnico.ulisboa.pt>

tiago-firmino commented:
pre-commit.ci autofix

andremcorreia committed:
Fix: pre-commit

rhshadrach reviewed (changes requested):
Only took a quick look, but overall this is looking good. Can you also add tests for EAs (the nullable and pyarrow dtypes).

tiago-firmino commented on pandas/core/_numba/executor.py:
```diff
@@ -69,13 +69,20 @@ def column_looper(
+                    labels,
+                    ngroups,
+                    min_periods,
+                    *args,
+                    skipna,
```
Okay, we will apply that change and update the kernel files accordingly.

andremcorreia commented on pandas/_libs/groupby.pyx:
```diff
@@ -733,32 +740,39 @@ def group_sum(
+                    else:
+                        sumx[lab, j] = val
+                        compensation[lab, j] = 0
+                        break
```
Correct, it was an oversight on our part.

tiago-firmino commented on pandas/tests/groupby/test_numba.py:
```diff
@@ -71,3 +72,57 @@ def test_no_engine_doesnt_raise(self):
+        df = DataFrame(
+            {
+                "l": ["A", "A", "A", "B", "B", "B"],
+                "int": [-1, 1, -1, 1, 1, np.nan],
```
Alright, we will remove the integer column.

andremcorreia commented on pandas/tests/groupby/test_reductions.py:
```diff
@@ -1047,6 +1047,120 @@ def scipy_sem(*args, **kwargs):
+)
+def test_skipna_reduction_ops_consistency(reduction_method):
+    # GH15675
+    # Testing if provinding skipna=True maintains the default functionality
```
Ok, we were a bit in doubt on whether to add those or not, will remove them then.

tiago-firmino commented on pandas/tests/groupby/test_reductions.py:
```diff
@@ -1192,3 +1306,31 @@ def test_groupby_std_datetimelike():
     tm.assert_frame_equal(result, expected)
+
+
+def test_skipna_string_sum():
```
Sure, we will do that.

tiago-firmino commented on pandas/core/resample.py:
```diff
@@ -1194,8 +1194,12 @@ def last(
     @doc(GroupBy.median)
-    def median(self, numeric_only: bool = False):
-        return self._downsample("median", numeric_only=numeric_only)
+    def median(
```
I apologize, it was an oversight on our part, we will take care of the docs for the remaining methods.

andremcorreia commented on pandas/tests/groupby/test_reductions.py:
```diff
@@ -1047,6 +1047,120 @@ def scipy_sem(*args, **kwargs):
+        {
+            "l": ["A", "A", "A", "A", "B", "B", "B", "B"],
+            "int": [-1, 1, -1, 2, 1, 1, 1, np.nan],
+            "float": [-1.0, 1.2, -1.1, 1.5, -1.1, 1.5, np.nan, 1.0],
```
Edited it, although I'm not entirely sure if we made the right parametrization. We will apply any further changes needed.

rhshadrach replied:
Sorry - I should have been more clear. Each assert only uses a single column. So I think you can do:

```python
@pytest.mark.parametrize(
    "reduction_method, values",
    [
        ("sum", [-1.0, 1.2, -1.1, 1.5, -1.1, 1.5, np.nan, 1.0]),
        ("sum", ["foo", "bar", "baz", "foo", "foo", "foo", pd.NA, "foo"]),
    ],
)
def test_skipna_reduction_ops_cython(reduction_method, values):
    df = pd.DataFrame({"key": [1, 1, ...], "values": values})
    gb = df.groupby("key")
    result = getattr(gb, reduction_method)(skipna=False)
    ...
```
This makes it easy to expand on if necessary, and only one test per parametrization (rather than looping over columns).

andremcorreia replied:
Understood, this does make the parametrization take quite a few lines but I agree that it's easier to expand and seems more consistent with that's done in other tests.

andremcorreia committed:
Reworked sugestions

Co-authored-by: Tiago Firmino <tiago.esteves.firmino@tecnico.ulisboa.pt>

tiago-firmino committed:
Reworked documentation

Co-authored-by: André Correia <andre.m.correia@tecnico.ulisboa.pt>

andremcorreia commented on pandas/tests/groupby/test_reductions.py:
```diff
@@ -1047,6 +1047,120 @@ def scipy_sem(*args, **kwargs):
+
+@pytest.mark.parametrize(
+    "reduction_method",
+    ["sum", "min", "max", "mean", "median", "prod", "sem", "std", "var"],
```
We added any, all and skew as those operations seem to work in the same way as the ones we added skipna to. I think the rest can be added over time.

andremcorreia commented on pandas/core/_numba/kernels/sum_.py:
```diff
@@ -232,7 +241,7 @@ def grouped_sum(
         prev_value = prev_vals[lab]
         sum_x = output[lab]
-        if nobs >= min_periods:
+        if not np.isnan(sum_x) and nobs >= min_periods:
```
Correct, but I'm not sure how that impacts the highlighted condition as it only checks the nobs if sum_x is not np.nan. Could you explain a bit further if possible?

tiago-firmino committed:
FIX: resample redefinition

Co-authored-by: André Correia <andre.m.correia@tecnico.ulisboa.pt>

andremcorreia committed:
FIX: Small tweaks in docs

Co-authored-by: Tiago Firmino <tiago.esteves.firmino@tecnico.ulisboa.pt>

andremcorreia committed:
Refactored test parameterization

Co-authored-by: Tiago Firmino <tiago.esteves.firmino@tecnico.ulisboa.pt>

tiago-firmino committed:
Added tests for EAs

Co-authored-by: André Correia <andre.m.correia@tecnico.ulisboa.pt>

andremcorreia committed:
Removed Arrow support

andremcorreia commented:
Hi, we refactored the tests as requested and were working on adding the EAs. We found and fixed a few sneaky edge case bugs with these, but we ran into a problem with ```dtype=pd.ArrowDtype(pa.int64())```.

Since the current implementation with arrows requires creating a typed numpy array and no NA value can be directly used in place for integers, we don't really see a path forward without considerable changes for this particular dtype.

We could add arrow floats easily, but we felt like only having a few arrow dtypes supported doesn't make much sense. 

We could try a while longer, and maybe there's an easier way to solve this that we aren't seeing, but we think it makes more sense to create a new issue for them and add them in the in the future. Specially since neither of us is that knowledgeable about pyarrows.

Changing seemingly required a considerable amount of other changes but the problems steems from this method.

pandas/core/arrays/arrow/array.py
```
    def _to_masked(self):
        pa_dtype = self._pa_array.type

        if pa.types.is_floating(pa_dtype) or pa.types.is_integer(pa_dtype):
            na_value = 1
        elif pa.types.is_boolean(pa_dtype):
            na_value = True
        else:
            raise NotImplementedError

        dtype = _arrow_dtype_mapping()[pa_dtype]
        mask = self.isna()
        arr = self.to_numpy(dtype=dtype.numpy_dtype, na_value=na_value)
        return dtype.construct_array_type()(arr, mask)
```

How should we proceed?

@rhshadrach

andremcorreia commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

andremcorreia committed:
pre-commit fix

andremcorreia committed:
Merge branch 'add_skipna_on_groupby_ops_pr' of github.com:andremcorreia/pandas into add_skipna_on_groupby_ops_pr

rhshadrach commented on pandas/core/groupby/ops.py:
```diff
@@ -392,6 +392,13 @@ def _call_cython_op(
                         values[mask] = True
             values = values.astype(bool, copy=False).view(np.int8)
             is_numeric = True
+        elif (
+            self.how in ["median", "sem", "std", "var"]
+            and "skipna" in kwargs
+            and not kwargs["skipna"]
+        ):
+            # if skipna=False we don't want to use masks created for Nullable dtypes
+            mask = None
```
The masks indicate where NA values are, don't we need to use them?

andremcorreia replied:
Correct, we had misunderstood how the masks were being used.

rhshadrach commented:
> Since the current implementation with arrows requires creating a typed numpy array and no NA value can be directly used in place for integers, we don't really see a path forward without considerable changes for this particular dtype.

The groupby methods implemented in Cython use `mask` and `result_mask` arguments for this purpose. This indicates where NA values are, so that they don't need to be in the NumPy array. If this isn't clear I can flesh it out some more, just ask.

rhshadrach reviewed (commented):
Once EAs are implemented, we'll also want a test for them in `tests/extension/base/reduce.py`.

tiago-firmino committed:
WIP EAs support

Co-authored-by: André Correia <andre.m.correia@tecnico.ulisboa.pt>

tiago-firmino commented:
Hello, after some rethinking about how we were handling EAs, we believe we've got it done correctly now since we were overthinking it before. However, we noticed the relevant array types tend to use ```pd.NA``` whilst the Series variants we are using to compare for our expected values are inferring ```np.nan```. 
We would like to know what to do in this situation, that is, what should be the expected value in EAs case, np.nan or pd.NA.

Our way of testing it was as such:
````

df = DataFrame({"key": [1, 1, 1, 2, 2, 2], "values": Series(pd.array([-1.0, 1.2, -1.1, 1.5, np.nan, 1.0], dtype="Float64"))})
gb = df.groupby("key")
result_cython = getattr(gb, reduction_method)(skipna=False)
expected = gb.apply(
        lambda x: getattr(x, reduction_method)(skipna=False), include_groups=False
)
tm.assert_frame_equal(result_cython, expected, check_exact=False, check_dtype=False)
`````
And the ```AssertionError``` we get is:
```
[-0.9000000000000001, <NA>]
E   Length: 2, dtype: Float64
E   [right]: [-0.9000000000000001, nan]
E   At positional index 1, first diff: <NA> != nan
```

In relation to the tests in ```tests/extension/base/reduce.py```, we would appreciate if some help with understanding how the tests work, how they are called and where the arguments come from. We understood they are the foundation for other tests used in the ```tests/extension``` directory but some fields left us confused.

rhshadrach commented:
> [tiago-firmino](https://github.com/tiago-firmino) [force-pushed](https://github.com/pandas-dev/pandas/compare/558bc25b1d3ed9d96537f78413948526ac8883fd..2856c6dd1a6451423bdcaea6ed2e92a325f84168) the add_skipna_on_groupby_ops_pr branch from[ 558bc25 ](https://github.com/pandas-dev/pandas/commit/558bc25b1d3ed9d96537f78413948526ac8883fd)to 2856c6d

@tiago-firmino - each time you force push, reviewers must review the entire PR as history could be changing. On small PRs, this isn't much of a problem, but this is not a small PR. Can I ask that you no longer force push here?

rhshadrach commented:
> However, we noticed the relevant array types tend to use `pd.NA` whilst the Series variants we are using to compare for our expected values are inferring `np.nan`.

Indeed, `apply` here looks to be producing the wrong result.

> In relation to the tests in `tests/extension/base/reduce.py`, we would appreciate if some help with understanding how the tests work, how they are called and where the arguments come from.

For this PR, the relevant tests are in `tests/extension/base/reduce.py` in the class `BaseReduceTests`. In `tests/extension/base/__init__.py`, the class `ExtensionTests` is then a subclass of this, so inherits its methods. pytest runs the files e.g. `tests/extension/test_arrow.py` where there is the class `TestArrowArray(base.ExtensionTests)`. This class has all the methods of `BaseReduceTests`, so the tests are run. Within `test_arrow.py` are fixtures such as `data` which setup data specifically for the pyarrow tests. Likewise, in `test_masked.py`, the fixture `data` sets up data specifically for the NumPy-nullable tests.

tiago-firmino commented:
> > [tiago-firmino](https://github.com/tiago-firmino) [force-pushed](https://github.com/pandas-dev/pandas/compare/558bc25b1d3ed9d96537f78413948526ac8883fd..2856c6dd1a6451423bdcaea6ed2e92a325f84168) the add_skipna_on_groupby_ops_pr branch from[ 558bc25 ](https://github.com/pandas-dev/pandas/commit/558bc25b1d3ed9d96537f78413948526ac8883fd)to 2856c6d
> 
> @tiago-firmino - each time you force push, reviewers must review the entire PR as history could be changing. On small PRs, this isn't much of a problem, but this is not a small PR. Can I ask that you no longer force push here?

I'm really sorry, I was not aware of that and will no longer do it.

andremcorreia committed:
Extension Array Support Tests

Co-authored-by: Tiago Firmino <tiago.esteves.firmino@tecnico.ulisboa.pt>

andremcorreia committed:
Merge branch 'main' into add_skipna_on_groupby_ops_pr

andremcorreia committed:
WIP: Fixing Tests

Co-authored-by: Tiago Firmino <tiago.esteves.firmino@tecnico.ulisboa.pt>

andremcorreia committed:
WIP: 32bit fix

andremcorreia commented:
Hello,
We've had a bit less free time lately, hence the delay, but we've run into some problems that we're not sure how to address. 
The main issue is with the prod function using int8, int16, and int32. Locally, we can't replicate the errors that show up in the pipeline, but we were able to identify that the wrong value is in the expected output, not in our result. However, after spending a long time looking at this issue, neither of us has been able to find a solution. 
Do you have any suggestions on how we could tackle this?

rhshadrach commented:
@andremcorreia - I should be able to take a look in the next few days.

rhshadrach commented:
@andremcorreia 

> but we were able to identify that the wrong value is in the expected output, not in our result.

It looks to me like the `result` is overflowing in 64-bit precision, the expected in 32-bit precision. On Windows, this will be fixed by NumPy 2.0. In any case, this isn't an issue with this PR. Can you add the following to the corresponding test:

```python
if op_name == "prod" and skipna and data.dtype.itemsize < 8 and np.intp().itemsize < 8:
    pytest.xfail(reason=f"{op_name} with itemsize {data.dtype.itemsize} overflows")
```

tiago-firmino committed:
WIP: overflow

Co-authored-by: André Correia <andre.m.correia@tecnico.ulisboa.pt>

andremcorreia committed:
Merge branch 'main' into add_skipna_on_groupby_ops_pr

andremcorreia committed:
Fix tests 32bit

andremcorreia committed:
small tweaks

andremcorreia committed:
simpler test skipping approach

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
